File: apps/api/src/country/__tests__/resolution-chain.test.ts
/**
 * Phase 1 — Country Resolution Chain unit tests.
 * Tests all 5 resolution steps + feature-flag OFF backward-compatibility.
 */

import { describe, it, expect } from 'vitest';
import type { FastifyRequest } from 'fastify';
import {
  resolveCountryFromRequest,
  resolveCountry,
} from '../resolution-chain.js';
import { GLOBAL_PROFILE, INDIA_PROFILE } from '../types.js';

/** Minimal FastifyRequest mock */
function req(headers: Record<string, string>): FastifyRequest {
  return { headers, log: { debug: () => {} } } as unknown as FastifyRequest;
}

describe('Country resolution chain', () => {
  describe('Step 1 — x-user-country header', () => {
    it('resolves known Tier-1 country from x-user-country', () => {
      const r = resolveCountryFromRequest(req({ 'x-user-country': 'US' }));
      expect(r.profile.isoCode).toBe('US');
      expect(r.resolvedBy).toBe('x-user-country');
    });

    it('resolves Tier-2 country from x-user-country', () => {
      const r = resolveCountryFromRequest(req({ 'x-user-country': 'JP' }));
      expect(r.profile.isoCode).toBe('JP');
    });

    it('falls through for unknown ISO code in x-user-country', () => {
      const r = resolveCountryFromRequest(req({ 'x-user-country': 'ZZ' }));
      expect(r.resolvedBy).not.toBe('x-user-country');
    });

    it('is case-insensitive', () => {
      const r = resolveCountryFromRequest(req({ 'x-user-country': 'de' }));
      expect(r.profile.isoCode).toBe('DE');
    });
  });

  describe('Step 2 — Accept-Language region', () => {
    it('extracts region from simple locale', () => {
      const r = resolveCountryFromRequest(req({ 'accept-language': 'en-GB' }));
      expect(r.profile.isoCode).toBe('GB');
      expect(r.resolvedBy).toBe('accept-language');
    });

    it('uses first tag in multi-value header', () => {
      const r = resolveCountryFromRequest(req({
        'accept-language': 'en-AU,en;q=0.9',
      }));
      expect(r.profile.isoCode).toBe('AU');
    });

    it('falls through when region not in registry', () => {
      const r = resolveCountryFromRequest(req({ 'accept-language': 'en-ZZ' }));
      expect(r.resolvedBy).not.toBe('accept-language');
    });

    it('falls through when locale has no region subtag', () => {
      const r = resolveCountryFromRequest(req({ 'accept-language': 'en' }));
      expect(r.resolvedBy).not.toBe('accept-language');
    });
  });

  describe('Step 3 — CF-IPCountry header', () => {
    it('uses Cloudflare country header', () => {
      const r = resolveCountryFromRequest(req({ 'cf-ipcountry': 'CA' }));
      expect(r.profile.isoCode).toBe('CA');
      expect(r.resolvedBy).toBe('cf-ipcountry');
    });

    it('ignores CF sentinel value XX (unknown IP)', () => {
      const r = resolveCountryFromRequest(req({ 'cf-ipcountry': 'XX' }));
      expect(r.resolvedBy).not.toBe('cf-ipcountry');
    });

    it('ignores CF sentinel value T1 (Tor)', () => {
      const r = resolveCountryFromRequest(req({ 'cf-ipcountry': 'T1' }));
      expect(r.resolvedBy).not.toBe('cf-ipcountry');
    });
  });

  describe('Step 4 — x-country-code header', () => {
    it('uses custom gateway header', () => {
      const r = resolveCountryFromRequest(req({ 'x-country-code': 'SG' }));
      expect(r.profile.isoCode).toBe('SG');
      expect(r.resolvedBy).toBe('x-country-code');
    });
  });

  describe('Step 5 — GLOBAL fallback', () => {
    it('returns GLOBAL when no headers present', () => {
      const r = resolveCountryFromRequest(req({}));
      expect(r.profile).toEqual(GLOBAL_PROFILE);
      expect(r.resolvedBy).toBe('fallback');
    });

    it('returns GLOBAL when all headers contain unknown codes', () => {
      const r = resolveCountryFromRequest(req({
        'x-user-country': 'ZZ',
        'accept-language': 'en',
        'cf-ipcountry': 'XX',
        'x-country-code': '??',
      }));
      expect(r.resolvedBy).toBe('fallback');
    });
  });

  describe('Priority ordering', () => {
    it('x-user-country wins over accept-language and cf-ipcountry', () => {
      const r = resolveCountryFromRequest(req({
        'x-user-country':  'US',
        'accept-language': 'en-GB',
        'cf-ipcountry':    'AU',
      }));
      expect(r.profile.isoCode).toBe('US');
      expect(r.resolvedBy).toBe('x-user-country');
    });

    it('accept-language wins over cf-ipcountry', () => {
      const r = resolveCountryFromRequest(req({
        'accept-language': 'en-IN',
        'cf-ipcountry':    'AU',
      }));
      expect(r.profile.isoCode).toBe('IN');
      expect(r.resolvedBy).toBe('accept-language');
    });
  });

  describe('Feature flag backward compat', () => {
    it('returns INDIA_PROFILE when flag is OFF regardless of headers', () => {
      const profile = resolveCountry(
        req({ 'x-user-country': 'US', 'cf-ipcountry': 'US' }),
        false, // flag OFF
      );
      expect(profile).toEqual(INDIA_PROFILE);
      expect(profile.isoCode).toBe('IN');
    });

    it('runs chain when flag is ON', () => {
      const profile = resolveCountry(
        req({ 'x-user-country': 'AE' }),
        true, // flag ON
      );
      expect(profile.isoCode).toBe('AE');
    });
  });

  describe('CountryProfile shape', () => {
    it('resolved profile has all required fields', () => {
      const r = resolveCountryFromRequest(req({ 'x-user-country': 'DE' }));
      const p = r.profile;
      expect(typeof p.isoCode).toBe('string');
      expect(typeof p.tier).toBe('string');
      expect(typeof p.displayName).toBe('string');
      expect(typeof p.locale).toBe('string');
      expect(typeof p.currencyCode).toBe('string');
      expect(typeof p.rtl).toBe('boolean');
      expect(typeof p.allergenRegime).toBe('string');
      expect(typeof p.nutritionStandard).toBe('string');
      expect(Array.isArray(p.mccList)).toBe(true);
    });

    it('RTL is true for AE (Arabic)', () => {
      const r = resolveCountryFromRequest(req({ 'x-user-country': 'AE' }));
      expect(r.profile.rtl).toBe(true);
    });

    it('RTL is false for India', () => {
      expect(INDIA_PROFILE.rtl).toBe(false);
    });

    it('India allergen regime is FSSAI_8', () => {
      expect(INDIA_PROFILE.allergenRegime).toBe('FSSAI_8');
    });

    it('US allergen regime is FDA_9_SESAME', () => {
      const r = resolveCountryFromRequest(req({ 'x-user-country': 'US' }));
      expect(r.profile.allergenRegime).toBe('FDA_9_SESAME');
    });

    it('EU countries allergen regime is EU_14', () => {
      for (const code of ['GB', 'DE', 'FR', 'IT', 'ES', 'NL']) {
        const r = resolveCountryFromRequest(req({ 'x-user-country': code }));
        expect(r.profile.allergenRegime).toBe('EU_14');
      }
    });
  });

  describe('Registry completeness', () => {
    it('all 8 Tier-1 countries resolve', () => {
      for (const code of ['IN', 'US', 'GB', 'AE', 'SG', 'AU', 'CA', 'DE']) {
        const r = resolveCountryFromRequest(req({ 'x-user-country': code }));
        expect(r.profile.tier).toBe('tier1');
        expect(r.profile.isoCode).toBe(code);
      }
    });

    it('all 17 Tier-2 countries resolve', () => {
      for (const code of ['JP','FR','KR','BR','MX','ID','TH','MY','PH','VN','ZA','NG','EG','SA','IT','ES','NL']) {
        const r = resolveCountryFromRequest(req({ 'x-user-country': code }));
        expect(r.profile.tier).toBe('tier2');
        expect(r.profile.isoCode).toBe(code);
      }
    });
  });
});

File: packages/country_engine/test/resolution_chain_test.dart
// Country engine — resolution chain + registry tests.

import 'package:flutter_test/flutter_test.dart';
import 'package:nutrimind_country_engine/nutrimind_country_engine.dart';

void main() {
  group('CountryRegistry', () {
    test('all 8 Tier-1 countries resolve', () {
      for (final code in ['IN', 'US', 'GB', 'AE', 'SG', 'AU', 'CA', 'DE']) {
        final p = CountryRegistry.lookup(code);
        expect(p, isNotNull, reason: '$code not found');
        expect(p!.tier, CountryTier.tier1);
      }
    });

    test('all 17 Tier-2 countries resolve', () {
      for (final code in [
        'JP','FR','KR','BR','MX','ID','TH','MY','PH','VN','ZA','NG','EG','SA','IT','ES','NL'
      ]) {
        final p = CountryRegistry.lookup(code);
        expect(p, isNotNull, reason: '$code not found');
        expect(p!.tier, CountryTier.tier2);
      }
    });

    test('case-insensitive lookup', () {
      expect(CountryRegistry.lookup('in')?.isoCode, 'IN');
      expect(CountryRegistry.lookup('us')?.isoCode, 'US');
    });

    test('unknown country returns null', () {
      expect(CountryRegistry.lookup('ZZ'), isNull);
    });

    test('lookupOrGlobal returns GLOBAL for unknown', () {
      final p = CountryRegistry.lookupOrGlobal('ZZ');
      expect(p.isoCode, 'GLOBAL');
      expect(p.tier, CountryTier.fallback);
    });

    test('MCC lookup works for Tier-1 countries', () {
      expect(CountryRegistry.lookupByMcc('404')?.isoCode, 'IN');
      expect(CountryRegistry.lookupByMcc('405')?.isoCode, 'IN');
      expect(CountryRegistry.lookupByMcc('310')?.isoCode, 'US');
      expect(CountryRegistry.lookupByMcc('234')?.isoCode, 'GB');
      expect(CountryRegistry.lookupByMcc('525')?.isoCode, 'SG');
    });

    test('unknown MCC returns null', () {
      expect(CountryRegistry.lookupByMcc('999'), isNull);
    });

    test('all profiles have non-empty locale and callingCode', () {
      for (final p in CountryRegistry.all) {
        expect(p.locale, isNotEmpty, reason: '${p.isoCode} locale empty');
        expect(p.callingCode, isNotEmpty, reason: '${p.isoCode} callingCode empty');
      }
    });
  });

  group('CountryProfile fields', () {
    test('India allergen regime is FSSAI_8', () {
      expect(CountryProfile.india.allergenRegime, AllergenRegime.fssai8);
    });

    test('India nutrition standard is ICMR-NIN', () {
      expect(CountryProfile.india.nutritionStandard, NutritionStandard.icmrNin);
    });

    test('US allergen regime is FDA_9_SESAME', () {
      final us = CountryRegistry.lookup('US')!;
      expect(us.allergenRegime, AllergenRegime.fda9Sesame);
    });

    test('EU countries allergen regime is EU_14', () {
      for (final code in ['GB', 'DE', 'FR', 'IT', 'ES', 'NL']) {
        expect(CountryRegistry.lookup(code)!.allergenRegime, AllergenRegime.eu14,
            reason: '$code should be EU_14');
      }
    });

    test('Japan allergen regime is JP_8', () {
      expect(CountryRegistry.lookup('JP')!.allergenRegime, AllergenRegime.jp8);
    });

    test('RTL is true for AE, EG, SA', () {
      for (final code in ['AE', 'EG', 'SA']) {
        expect(CountryRegistry.lookup(code)!.rtl, isTrue, reason: '$code should be RTL');
      }
    });

    test('RTL is false for IN, US, GB, JP', () {
      for (final code in ['IN', 'US', 'GB', 'JP']) {
        expect(CountryRegistry.lookup(code)!.rtl, isFalse, reason: '$code should not be RTL');
      }
    });

    test('JSON round-trip preserves all fields', () {
      final original = CountryProfile.india;
      final decoded  = CountryProfile.fromJson(original.toJson());
      expect(decoded.isoCode,          original.isoCode);
      expect(decoded.tier,             original.tier);
      expect(decoded.allergenRegime,   original.allergenRegime);
      expect(decoded.nutritionStandard, original.nutritionStandard);
      expect(decoded.rtl,              original.rtl);
      expect(decoded.mccList,          original.mccList);
    });

    test('toJson emits the real server wire-format strings, not Dart enum member names', () {
      // Phase 10 regression: fromJson()/toJson() used to round-trip via `.name`/`byName()`,
      // which only ever matched Dart's own enum member names ('fssai8', 'icmrNin') — never the
      // real values apps/api/src/country/types.ts actually sends ('FSSAI_8', 'ICMR_NIN'). This
      // was undetected because the only prior caller was a self-consistent SharedPreferences
      // round-trip (Dart writing then reading its own output), never real API JSON.
      final json = CountryProfile.india.toJson();
      expect(json['allergenRegime'], 'FSSAI_8');
      expect(json['nutritionStandard'], 'ICMR_NIN');
    });

    test('fromJson parses the real server wire-format strings correctly', () {
      final json = {
        'isoCode': 'GB', 'tier': 'tier1', 'displayName': 'United Kingdom',
        'locale': 'en_GB', 'currencyCode': 'GBP', 'rtl': false,
        'allergenRegime': 'EU_14', 'nutritionStandard': 'UK_SACN',
        'callingCode': '+44', 'mccList': ['234', '235'],
      };
      final profile = CountryProfile.fromJson(json);
      expect(profile.allergenRegime, AllergenRegime.eu14);
      expect(profile.nutritionStandard, NutritionStandard.ukSacn);
    });
  });

  group('CountryResolutionChain', () {
    const chain = CountryResolutionChain();

    test('Step 1: stored override wins', () async {
      final r = await chain.resolve(storedOverride: 'US');
      expect(r.profile.isoCode, 'US');
      expect(r.resolvedBy, 'stored-override');
    });

    test('Step 2: API profile wins when no override', () async {
      final r = await chain.resolve(apiProfileCountry: 'GB');
      expect(r.profile.isoCode, 'GB');
      expect(r.resolvedBy, 'api-profile');
    });

    test('Step 1 wins over Step 2', () async {
      final r = await chain.resolve(storedOverride: 'CA', apiProfileCountry: 'AU');
      expect(r.profile.isoCode, 'CA');
    });

    test('Step 4: OS locale region', () async {
      final r = await chain.resolve(osLocale: 'de_DE');
      expect(r.profile.isoCode, 'DE');
      expect(r.resolvedBy, 'os-locale');
    });

    test('Step 5: stored last-known fallback', () async {
      final r = await chain.resolve(storedLastKnown: 'SG');
      expect(r.profile.isoCode, 'SG');
      expect(r.resolvedBy, 'stored-last-known');
    });

    test('Step 6: GLOBAL when all steps fail', () async {
      final r = await chain.resolve(
        storedOverride: null,
        apiProfileCountry: null,
        osLocale: 'zh',  // no region subtag
        storedLastKnown: null,
      );
      expect(r.profile.isoCode, 'GLOBAL');
      expect(r.resolvedBy, 'fallback');
    });

    test('unknown override falls through to next step', () async {
      final r = await chain.resolve(storedOverride: 'ZZ', storedLastKnown: 'AU');
      expect(r.profile.isoCode, 'AU');
    });

    test('locale with unknown region falls through', () async {
      final r = await chain.resolve(osLocale: 'en_ZZ', storedLastKnown: 'NZ');
      // NZ is not in registry → GLOBAL
      expect(r.resolvedBy, 'fallback');
    });
  });

  group('regionFromLocale', () {
    test('extracts region from BCP-47 dash form', () {
      expect(regionFromLocale('en-US'), 'US');
      expect(regionFromLocale('en-IN'), 'IN');
      expect(regionFromLocale('ar-AE'), 'AE');
    });

    test('extracts region from underscore form', () {
      expect(regionFromLocale('en_GB'), 'GB');
      expect(regionFromLocale('de_DE'), 'DE');
    });

    test('returns null for bare language code', () {
      expect(regionFromLocale('en'), isNull);
      expect(regionFromLocale('hi'), isNull);
    });
  });
}
